### ISSUE TITLE
Slow Reading

### ISSUE THREAD

**Issue by classicboss302** [JoshClose/CsvHelper#412]
I have a very large CSV file in term of columns. The performance of CsvHelper seems to degrade quickly in this scenario. The Gist below is for an example CSV file (just the headers and the first row of data). It takes around 8 seconds to read those two rows in for me. That goes downhill quickly as I use a full file.

Is there some setting I can change to help with this? I tried changing the buffer size and that didn't really do much.

I am reading each line with the parser.

parser.Read()

https://gist.githubusercontent.com/classicboss302/8c4dbb52bb6a4175cbac/raw/b6439c43f62b2896ffd240ad1eb5ad8487fc785d/SlowCsv

**Comment by JoshClose** [comment]
Wow, that is insanely slow. It should be pretty much instant with that small data.

I will look into what is causing the slow down.

**Comment by Orcomp** [comment]
I am going off memory here (so I might be completely wrong...). We found some performance degradation when matching a column name to a column index a few months back.

I think CsvHelper would try and resolve the column index for each cell it is trying to parse (or something like that.)

We managed to get rid of the performance degradation (when dealing with a lot of column headers) by creating a dictionary mapping the column names to column index.

If you ignore column headers, does it import quickly?

**Comment by Orcomp** [comment]
Interestingly even Excel cannot open the file...

**Comment by classicboss302** [comment]
Setting HasHeaderRecord to false doesn't give any speed improvements.

**Comment by JoshClose** [comment]
It can do a "normal" file of 500 MB faster than it can do this 5 MB file... I recently installed Win10 and am having some trouble getting debugging working in VS2015 for whatever reason, so I won't be able to test this for a bit here.

My guess without being able to test it is the array resizing is causing it. https://github.com/JoshClose/CsvHelper/blob/master/src/CsvHelper/CsvParser.cs#L197

Can you try changing that size the exact number of fields there are and see if that helps? I should be able to fix that pretty easily if that is the case.

Either download and edit the source, or use reflection to edit it.

Here is an example of how you can set the `FieldCount` with reflection. This is off the top of my head, so it may need to be changed to work properly. :)

```cs
var csv = new CsvParser( textReader );
csv.GetType().GetProperty( "FieldCount" ).SetValue( csv, 16384 );
```

**Comment by classicboss302** [comment]
I did some quick profiling and here are the hotspots for me. Definitely the resize as you suggested is part of the issue. For the other, I think switching to a StringBuilder would be better then using a String when you are modifying it.

Setting the number of fields removed AddFieldToRecord as a hotspot. Now it is just GetChar.

![addfieldtorecord](https://cloud.githubusercontent.com/assets/12926413/9200487/a46287b0-400d-11e5-9820-f4c6eb16ff43.png)
![getchar](https://cloud.githubusercontent.com/assets/12926413/9200486/a461b47a-400d-11e5-9aeb-d4023172e26e.png)

**Comment by JoshClose** [comment]
I think the string portion is as fast as it can be, but test it out if you like. I'd like to see what your findings are.

What I found is that taking a buffer and just copying the contents of the buffer onto the string was the fastest possible way to do the concatenation.

**Comment by classicboss302** [comment]
I tested switching from String to StringBuilder in CsvParser and it does show a good improvement.

Here are the results for a different test file.
* Without StringBuilder or setting the field count: 15.8 seconds
* With StringBuilder, but without setting the field count: 10.8 seconds
* Without StringBuilder, but with setting the field count: 10.5 seconds
* With StringBuilder and setting the field count: 2.7 seconds

**Comment by JoshClose** [comment]
I did a bunch of speed improvements to the parser, and this is one of them. It goes through your sample file in under a second. I changed the `field` and `rawRecord` to a `StringBuilder`. This actually slows down normal parsing a bit, but it's very minimal. Parsing a lot of columns is improved ridiculously though. I also created a custom type to handle the records that manages growing the array and efficiently adding new strings to it.

Here is the code, which will be doing away when this branch is merged. If I remember, I'll update these links to the files in master after merged.

The record manager. https://github.com/JoshClose/CsvHelper/blob/faster-parser/src/CsvHelper/RecordBuilder.cs

The new field reader that contains the StringBuilder. https://github.com/JoshClose/CsvHelper/blob/faster-parser/src/CsvHelper/FieldReader.cs
